test.js

```diff
@@ -244,3 +246,87 @@ test('sync (not found, custom cwd)', t => {
 
 	t.is(filePath, null);
 });
+
+test('async (custom function)', async t => {
```

Comment:
Can you add a test of what happens when the match function throws?

Reply:
Will do. On a related note, I realized recently that it looks like our async method is swallowing promise rejections, regardless of whether they come from `locate-path` or a custom matcher function. It is happening on `master` too, not specific to this PR. I will fix that while I'm here.